### ISSUE TITLE
Wav2Vec2.0 finetuning custom dataset using crossentropy not CTC, how to audio data preprocessing

### ISSUE THREAD

**Issue by qute012** [pytorch/fairseq#2917]
Thank you for great work.
I want to downstream pretrained model for keyword spotting.  Layers were added as follows:
```
self.w2v_encoder = task.build_model(cfg.model)
self.w2v_encoder.load_state_dict(state_dict)
        
self.avg_pool = nn.AdaptiveAvgPool1d(1)
self.classfier = nn.Linear(encoder_hidden_dim, self.n_class)
self.act = nn.ReLU(inplace=True)
```

when wav2vec encoder implements below custom dataset. Raw signals are transformed by a function of RawAudioDataset class in pairseq, and uses collater in that class as collat_fn in dataloader.
```
from fairseq.data.audio.raw_audio_dataset import * 

class SolarDataset(RawAudioDataset):

    def __init__(self, mode='training', root='/root/storage/dataset/solar_dataset/inter', sample_rate=16000):
        super(SolarDataset, self).__init__(
            sample_rate,
            pad=False
        )
        self.root = os.path.join(root,mode)
        self.sample_rate = sample_rate
        self.data = list()
        self.prep_dataset()
            
    def prep_dataset(self):
        for root, dir, files in os.walk(self.root):
            for file in files:
                f_path, cmd = os.path.join(root, file), root.split('/')[-1]
                self.data.append((f_path, cmd))

    def __getitem__(self, idx):
        f_path, cmd = self.data[idx]

        wav, curr_sample_rate = sf.read(f_path)
        if curr_sample_rate!=self.sample_rate:
            wav, curr_sample_rate = librosa.resample(wav, curr_sample_rate, self.sample_rate), self.sample_rate
        feats = torch.from_numpy(wav).float()
        feats = self.postprocess(feats, curr_sample_rate)
        y = CLASSES.index(cmd)
        #print(feats.shape)
        return {"id": y, "source": feats}
```
The problem is when implementing the w2v encoder, the tensor becomes Nan. I wonder if the audio preprocessing used in wav2vec 2.0 is the above method, and how to finetuning it using CrossEntropy, not CTC.

This is the value of tensor after passing the encoder.
![image](https://user-images.githubusercontent.com/33983084/99630826-db181800-2a7d-11eb-9e44-306bd33c1140.png)

Help me, Thank you in advance.

**Comment by alexeib** [comment]
few comments:
a) dont use id to store your target, it is supposed to be unique. just add another key like "y": ClASSSES.index(cmd) and then overwrite the collater to stack them. (assuming i interpreted your code correctly)
b) i would also maybe experiment with modifying the model to add a "cls" token to the beginning of the sequence and use that for classification. also where is that act() applied? before or after your linear classifier
c) it is not clear how you are doing finetuning now. in wav2vec_asr we just extract features from the pretrained model without doing the whole quantization step etc. there is a parameter you can pass, called features_only, i believe. you may also want to disable masking and things like that
d) nans in the output of feature encoder (e.g. look at features_pen) probably mean there is something wrong with the data you are feeding it. also you probably want to keep pad=True because otherwise the dataset will sample random crops of data

**Comment by qute012** [comment]
@alexeib Thanks for response!
I want to fintune the pretrained w2v encoder with keyword spotting first.

full model architectures as follows:
```
class KWS(nn.Module):
    def __init__(self, n_class=30, encoder_hidden_dim=101, cfg=None, state_dict=None):
        super(KWS, self).__init__()
        self.n_class = n_class
        assert not cfg is None
        assert not state_dict is None
        
        self.w2v_encoder = task.build_model(cfg.model)
        self.w2v_encoder.load_state_dict(state_dict)
        
        self.avg_pool = nn.AdaptiveAvgPool1d(1)
        self.classfier = nn.Linear(encoder_hidden_dim, self.n_class)
        #self.dropout = nn.Dropout(dropout_p)
        self.act = nn.ReLU(inplace=True)
        
    def forward(self, x):
        #print(x)
        output = self.w2v_encoder(**x)
        print(output)
        output=output['x']
        #CxBxT => BxCxT
        #print(output)
        c,b,t = output.shape
        output = output.view(b,c,t)
        output = self.avg_pool(output).view(b,c)
        #print(output)
        output = self.classfier(output)
        
        return self.act(output)
```
The output that passed w2v is the following shape(e. g. (101, B, T) => B:Batch Size, T:Time). Adaptive Pooling is used to connect the size of Channel and Time series with FC Layer.

b,c) Is it right to add cls token like Bert and then add a classifier, using the Feature_only parameter instead of Adaptive Pooling?
d) I wonder if input data of wav2vec 2.0 is correct to handle only the postprocess function of the RawAudioDataset class in the signal of wav file.

Thanks in advance.

**Comment by qute012** [comment]
Additionally, The model has been modified as follows, but the accuracy does not increase as the learning progresses.
```
class KWS(nn.Module):
    def __init__(self, n_class=30, encoder_hidden_dim=1024, cfg=None, state_dict=None):
        super(KWS, self).__init__()
        self.n_class = n_class
        assert not cfg is None
        assert not state_dict is None
        
        self.w2v_encoder = task.build_model(cfg.model)
        self.w2v_encoder.load_state_dict(state_dict)
        
        self.avg_pool = nn.AdaptiveAvgPool1d(1)
        self.classifier = nn.Linear(encoder_hidden_dim, self.n_class)
        #self.dropout = nn.Dropout(dropout_p)
        #self.act = nn.ReLU(inplace=True)
        
    def forward(self, x):
        #print(x)
        output = self.w2v_encoder(**x, features_only=True)
        #print(output)
        #print(output.shape)
        output=output['x']
        #print(output)
        #print(output.shape)
        #CxBxT => BxCxT
        #print(output)
        b,t,c = output.shape
        output = output.reshape(b,c,t)
        output = self.avg_pool(output).view(b,c)
        #print(output.shape)
        #print(output)
        output = self.classifier(output)
        
        return output
        
        
model = KWS(cfg=cfg, state_dict=state_dict['model']).cuda()
optimizer = torch.optim.Adam([
    {'params': model.w2v_encoder.parameters(), 'lr': 5e-5},
    {'params': model.classifier.parameters(), 'lr': 5e-4}
])
criterion = torch.nn.CrossEntropyLoss().cuda()
```

Loss doesn't decrease much either.
![image](https://user-images.githubusercontent.com/33983084/99813923-9e3a4700-2b8b-11eb-95f6-be38c1439d36.png)
